arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/raise/Raise.kt

```diff
@@ -331,69 +253,168 @@ public interface Raise<in R> {
+public inline fun <reified T : Throwable, E, A> recover(
+  @BuilderInference action: Raise<E>.() -> A,
+  @BuilderInference recover: (E) -> A,
+  @BuilderInference catch: (T) -> A,
+): A = fold(action, { t -> if (t is T) catch(t) else throw t }, recover, ::identity)
+
+/**
+ * Allows safely catching exceptions without capturing [CancellationException],
+ * or fatal exceptions like `OutOfMemoryError` or `VirtualMachineError` on the JVM.
+ *
+ * <!--- INCLUDE
+ * import arrow.core.Either
+ * import arrow.core.raise.either
+ * import arrow.core.raise.catch
+ * import io.kotest.matchers.shouldBe
+ * -->
+ * ```kotlin
+ * fun test() {
+ *   catch({ throw RuntimeException("BOOM") }) { t ->
+ *     "fallback"
+ *   } shouldBe "fallback"
+ *
+ *   fun fetchId(): Int = throw RuntimeException("BOOM")
+ *
+ *   either {
+ *     catch({ fetchId() }) { t ->
+ *       raise("something went wrong: ${t.message}")
+ *     }
+ *   } shouldBe Either.Left("something went wrong: BOOM")
+ * }
+ * ```
+ * <!--- KNIT example-raise-dsl-08.kt -->
+ * <!--- TEST lines.isEmpty() -->
+ *
+ * Alternatively, you can use `try { } catch { }` blocks with [nonFatalOrThrow].
+ * This API offers a similar syntax as the top-level [catch] functions like [Either.catch].
+ */
+@RaiseDSL
+public inline fun <A> catch(action: () -> A, catch: (Throwable) -> A): A =
+  try {
+    action()
+  } catch (t: Throwable) {
+    catch(t.nonFatalOrThrow())
   }
-  return fold({ action(this) }, { catch(it) }, { raise(it) }, { it })
-}
 
+/**
+ * Allows safely catching exceptions of type `T` without capturing [CancellationException],
+ * or fatal exceptions like `OutOfMemoryError` or `VirtualMachineError` on the JVM.
+ *
+ * <!--- INCLUDE
+ * import arrow.core.Either
+ * import arrow.core.raise.either
+ * import arrow.core.raise.catch
+ * import io.kotest.matchers.shouldBe
+ * -->
+ * ```kotlin
+ * fun test() {
+ *   catch({ throw RuntimeException("BOOM") }) { t ->
+ *     "fallback"
+ *   } shouldBe "fallback"
+ *
+ *   fun fetchId(): Int = throw RuntimeException("BOOM")
+ *
+ *   either {
+ *     catch({ fetchId() }) { t: RuntimeException ->
+ *       raise("something went wrong: ${t.message}")
+ *     }
+ *   } shouldBe Either.Left("something went wrong: BOOM")
+ * }
+ * ```
+ * <!--- KNIT example-raise-dsl-09.kt -->
+ * <!--- TEST lines.isEmpty() -->
+ *
+ * Alternatively, you can use `try { } catch(e: T) { }` blocks.
+ * This API offers a similar syntax as the top-level [catch] functions like [Either.catch].
+ */
 @RaiseDSL
 @JvmName("catchReified")
-public inline fun <reified T : Throwable, R, A> Raise<R>.catch(
-  @BuilderInference action: Raise<R>.() -> A,
-  @BuilderInference catch: Raise<R>.(T) -> A,
-): A {
-  contract {
-    callsInPlace(action, EXACTLY_ONCE)
-    callsInPlace(catch, AT_MOST_ONCE)
-  }
-  return catch(action) { t: Throwable -> if (t is T) catch(t) else throw t }
-}
+public inline fun <reified T : Throwable, A> catch(action: () -> A, catch: (T) -> A): A =
```

Comment:
This also matches what we used, and needed in [Gh Alerts Project](https://github.com/47deg/gh-alerts-subscriptions-kotlin/blob/858669890b96d9be3d0984e0d2cee7cbcc8110ab/src/main/kotlin/alerts/predef.kt#L23).

Wondering if this should be on `Raise` companion similar to `Either.catch` 🤔 They're basically the duality of each-other. Similarly, does `Either.catch` need a `reified` variant?

Reply:
> Raise sort of feels like much more of a primitive than Either
It absolutely is!

>  The name Raise.catch also just feels weird from a language standpoint
Thanks for the feedback ☺️ `catch` is also applicable outside of `Raise`, so that makes sense!